pallets/creditcoin/src/ocw/tasks/collect_coins.rs

```diff
@@ -913,14 +958,56 @@ pub(crate) mod tests {
 			let (_, to, ..) = generate_address_with_proof("collector");
 			let tx_id = &TX_HASH.hex_to_address();
 
-			let rpc_url = &CONTRACT_CHAIN.rpc_url().unwrap();
+			let rpc_url = &CHAIN.rpc_url().unwrap();
 			let mut tx = rpc::eth_get_transaction(tx_id, rpc_url).unwrap();
 			let tx_receipt = rpc::eth_get_transaction_receipt(tx_id, rpc_url).unwrap();
 			let eth_tip = rpc::eth_get_block_number(rpc_url).unwrap();
-			validate_collect_coins(&to, &tx_receipt, &tx, eth_tip).expect("valid");
+			let PassingCollectCoins { contract_address, .. } = Default::default();
+			validate_collect_coins(&to, &tx_receipt, &tx, eth_tip, &contract_address)
+				.expect("valid");
 			// Forged selector
-			tx.input.0[0] = 1;
-			validate_collect_coins(&to, &tx_receipt, &tx, eth_tip).expect_err("invalid");
+			tx.set_input(b"");
+			validate_collect_coins(&to, &tx_receipt, &tx, eth_tip, &contract_address)
+				.expect_err("invalid");
+		});
+	}
+
+	#[test]
+	fn set_collect_coins_only_as_root() {
+		let mut ext = ExtBuilder::default();
+		let acct_pubkey = ext.generate_authority();
+		let _auth = AccountId::from(acct_pubkey.into_account().0);
+		ext.build_and_execute(|| {
+			let contract = GCreContract::default();
+			assert_ok!(Creditcoin::<Test>::set_collect_coins_contract(
+				RawOrigin::Root.into(),
+				contract.clone()
+			));
+
+			let (acc, ..) = generate_address_with_proof("somebody");
+
+			assert_noop!(
+				Creditcoin::<Test>::set_collect_coins_contract(
+					RawOrigin::Signed(acc).into(),
+					contract.clone()
+				),
+				BadOrigin
+			);
+
+			assert_noop!(
+				Creditcoin::<Test>::set_collect_coins_contract(RawOrigin::None.into(), contract),
+				BadOrigin
+			);
+		});
+	}
+
+	#[test]
+	fn gcrecontract_value_query_is_default() {
+		let contract = GCreContract::default();
+		let ext = ExtBuilder::default();
+		ext.build_and_execute(|| {
+			let value_query = Creditcoin::<Test>::collect_coins_contract();
+			assert_eq!(contract, value_query);
```

Comment:
I can't figure out where this behavior comes from. If `set_collect_coins()` extrinsic hasn't been executed I expect storage to be empty. How come there's a default value? (or rather where exactly is that specified in the source? ). 

Also probably makes sense to use a non-default value for `contract` and then make sure it was written to storage. (lines 981-986 above).

Reply:
> I can't figure out where this behavior comes from. If set_collect_coins() extrinsic hasn't been executed I expect storage to be empty. How come there's a default value? (or rather where exactly is that specified in the source? ).

When you declare [storage](https://github.com/gluwa/creditcoin/pull/474/files#r934736783) with the macro you can set storage as either OptionQuery (implicit) or ValueQuery. ValueQuery uses the type's default if a factory method is not specified.

> Also probably makes sense to use a non-default value for contract and then make sure it was written to storage. (lines 981-986 above).

I agree, I missed that.

Reply:
#492 

> Also probably makes sense to use a non-default value for contract and then make sure it was written to storage. (lines 981-986 above).